python/python/tests/test_scalar_index.py

```diff
@@ -3503,3 +3503,150 @@ def test_btree_query_comparison_parametrized(
             f"Test '{test_name}' failed: Fragment index "
             f"and complete index returned different results for filter: {filter_expr}"
         )
+
+
+def test_fts_flat_fallback_matches_wand(tmp_path):
+    # Repro: when filter matches < 10%, FTS fell back to flat search and missed results.
+    # Two-term query increases reproduction likelihood.
+    # Compare default scan (prefilter=True) vs WAND-path (prefilter=False).
+
+    # Deterministic data
+    random.seed(123)
+    np.random.seed(123)
+
+    n = 2000
+    # 5% category 'A' to trigger fallback (default threshold 10%)
+    categories = np.where(np.random.rand(n) < 0.05, "A", "B")
+
+    vocab = [
+        "alpha",
+        "bravo",
+        "charlie",
+        "delta",
+        "echo",
+        "foxtrot",
+        "golf",
+        "hotel",
+        "india",
+    ]
+    needle1 = "needle"
+    needle2 = "pin"
+
+    texts = []
+    a_indices = [i for i, c in enumerate(categories) if c == "A"]
+    # Ensure many matches within the filtered set (both terms present)
+    a_with_needle = set(
+        random.sample(
+            a_indices,
+            max(1, len(a_indices) // 2),
+        )
+    )
+    for i in range(n):
+        toks = random.choices(vocab, k=random.randint(5, 12))
+        if i in a_with_needle:
+            # Add both terms for many A rows
+            toks.append(needle1)
+            toks.append(needle2)
+        # Also sprinkle some needles in B rows to make ranking realistic
+        elif categories[i] == "B" and random.random() < 0.03:
+            # Randomly add one of the terms to some B rows
+            toks.append(needle1 if random.random() < 0.5 else needle2)
+        texts.append(" ".join(toks))
+
+    tbl = pa.table(
+        {
+            "id": np.arange(n, dtype=np.int64),
+            "category": pa.array(categories.astype(str)),
+            "text": pa.array(texts, type=pa.large_string()),
+        }
+    )
+
+    ds_path = tmp_path / "fts_fallback.lance"
+    ds = lance.write_dataset(tbl, ds_path)
```

Comment:
I wonder if it would be useful to consider a case of multiple fragments dataset as well in that test ?